packages/shared_preferences/shared_preferences/lib/shared_preferences_async.dart

```diff
@@ -0,0 +1,370 @@
+/// missed changes that may have occurred since the cache was last updated.
+@immutable
+class SharedPreferencesWithCache {
+  /// Creates a new instance with the given options.
+  SharedPreferencesWithCache._create({
+    required this.sharedPreferencesOptions,
+    required SharedPreferencesWithCacheOptions cacheOptions,
+    Map<String, Object?>? cache,
+  })  : _cacheOptions = cacheOptions,
+        _directAccess =
+            SharedPreferencesAsync(options: sharedPreferencesOptions),
+        _cache = cache ?? <String, Object?>{};
+
+  /// Creates a new instance with the given options and reloads the cache from
+  /// the platform data.
+  static Future<SharedPreferencesWithCache> create({
+    required SharedPreferencesOptions sharedPreferencesOptions,
+    required SharedPreferencesWithCacheOptions cacheOptions,
+    Map<String, Object?>? cache,
+  }) async {
+    final SharedPreferencesWithCache preferences =
+        SharedPreferencesWithCache._create(
+      sharedPreferencesOptions: sharedPreferencesOptions,
+      cacheOptions: cacheOptions,
+      cache: cache,
+    );
+
+    await preferences.reloadCache();
+
+    return preferences;
+  }
+
+  /// Cache containing in-memory data.
+  final Map<String, Object?> _cache;
+
+  /// Options that determine the behavior of contained methods, usually
+  /// platform specific extensions of the [SharedPreferencesOptions] class.
+  final SharedPreferencesOptions sharedPreferencesOptions;
+
+  /// Options that define cache behavior.
+  final SharedPreferencesWithCacheOptions _cacheOptions;
+
+  /// Async access directly to the platform.
+  ///
+  /// Methods called through [_directAccess] will NOT update the cache.
+  final SharedPreferencesAsync _directAccess;
+
+  /// Updates cache with latest values from platform.
+  ///
+  /// This should be called before reading any values if the values may have
+  /// been changed by anything other than this cache instance,
+  /// such as from another isolate or native code that changes the underlying
+  /// preference storage directly.
+  Future<void> reloadCache() async {
+    _cache.clear();
+    _cache.addAll(await _directAccess
+        .getAll(GetPreferencesParameters(filter: _cacheOptions.filter)));
+  }
+
+  /// Returns true if cache contains the given [key].
+  bool containsKey(String key) {
+    if (!_isValidKey(key)) {
+      throw ArgumentError(
+          '$key is not included in the PreferencesFilter allowlist');
+    }
+    return _cache.containsKey(key);
+  }
+
+  /// Returns all keys in the cache.
+  Set<String> get keys {
+    return _cache.keys.toSet();
+  }
+
+  /// Reads a value of any type from the cache.
+  Object? get(String key) {
+    if (!_isValidKey(key)) {
+      throw ArgumentError(
+          '$key is not included in the PreferencesFilter allowlist');
+    }
+    return _cache[key];
+  }
+
+  /// Reads a value from the cache, throwing a [TypeError] if the value is not a
```

Comment:
If we want to specifically say it's a `[TypeError]`, then we need to actually make that a documented part of the contract of the platform interface API. Otherwise it could easily be a `PlatformException`, and if we don't want that (which we don't per [this](https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#platform-exception-handling)) we need to tell implementors that it's a requirement to transform that internally.

Reply:
And we should add platform implementation package integration tests on all the platforms that we actually do that correctly, if we don't already have that in this PR (I haven't checked), by having a test for each type that sets one type, then reads another.

Reply:
This is only reading from the dart cache, it shouldn't be reliant on platform error handling.

Reply:
Oops, let's pretend this comment is on line 54 😁